File: frontend/types/automation.ts
export type AutomationType = "scheduled_post" | "auto_reply" | "keyword_reply" | "scheduled_comment";
export type AutomationStatus = "active" | "paused" | "completed" | "failed";
export type PostType = "text" | "link";

export interface Automation {
  id: number;
  name: string;
  automation_type: AutomationType;
  status: AutomationStatus;
  subreddit?: string;
  scheduled_at?: string;
  cron_expression?: string;
  is_recurring: boolean;
  post_title?: string;
  post_content?: string;
  post_type: PostType;
  post_url?: string;
  reply_template?: string;
  trigger_keyword?: string;
  reply_to_post_id?: string;
  last_run_at?: string;
  next_run_at?: string;
  run_count: number;
  last_error?: string;
  created_at?: string;
  config?: Record<string, unknown>;
}

export const AUTOMATION_TYPE_LABELS: Record<AutomationType, string> = {
  scheduled_post: "Scheduled Post",
  auto_reply: "Auto Reply",
  keyword_reply: "Keyword Reply",
  scheduled_comment: "Scheduled Comment",
};

export const AUTOMATION_TYPE_DESCRIPTIONS: Record<AutomationType, string> = {
  scheduled_post: "Post to a subreddit at a specific time or on a recurring schedule",
  auto_reply: "Automatically reply to a specific Reddit post",
  keyword_reply: "Monitor a subreddit and reply to comments containing a keyword",
  scheduled_comment: "Post a comment on an existing post at a scheduled time",
};

export const STATUS_COLORS: Record<AutomationStatus, string> = {
  active: "text-green-400 bg-green-400/10",
  paused: "text-yellow-400 bg-yellow-400/10",
  completed: "text-blue-400 bg-blue-400/10",
  failed: "text-red-400 bg-red-400/10",
};

File: backend/models/automation.py
from sqlalchemy import Column, Integer, String, DateTime, Boolean, Text, ForeignKey, JSON
from sqlalchemy.orm import relationship
from database import Base
from datetime import datetime, timezone

class Automation(Base):
    __tablename__ = "automations"

    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False)

    name = Column(String, nullable=False)
    automation_type = Column(String, nullable=False)  # "scheduled_post", "auto_reply", "scheduled_comment", "upvote_monitor"
    status = Column(String, default="active")  # active, paused, completed, failed

    # Common fields
    subreddit = Column(String, nullable=True)

    # Scheduling
    scheduled_at = Column(DateTime, nullable=True)  # one-time
    cron_expression = Column(String, nullable=True)  # recurring
    is_recurring = Column(Boolean, default=False)

    # Post fields
    post_title = Column(String, nullable=True)
    post_content = Column(Text, nullable=True)
    post_type = Column(String, default="text")  # text, link, image
    post_url = Column(String, nullable=True)

    # Reply/comment fields
    reply_template = Column(Text, nullable=True)
    trigger_keyword = Column(String, nullable=True)
    reply_to_post_id = Column(String, nullable=True)

    # Execution tracking
    last_run_at = Column(DateTime, nullable=True)
    next_run_at = Column(DateTime, nullable=True)
    run_count = Column(Integer, default=0)
    last_error = Column(Text, nullable=True)

    # Extra config
    config = Column(JSON, nullable=True)

    created_at = Column(DateTime, default=lambda: datetime.now(timezone.utc))
    updated_at = Column(DateTime, default=lambda: datetime.now(timezone.utc), onupdate=lambda: datetime.now(timezone.utc))

    user = relationship("User", back_populates="automations")
